font/src/ft/mod.rs

```diff
@@ -529,6 +589,86 @@ impl FreeTypeRasterizer {
     }
 }
 
+fn downsample_bitmap(mut bitmap_glyph: RasterizedGlyph, fixup_factor: f64) -> RasterizedGlyph {
+    // Don't try to upscale
+    if fixup_factor > 1.0 {
+        return bitmap_glyph;
+    }
+
+    let bitmap_width = bitmap_glyph.width as f64;
+    let bitmap_height = bitmap_glyph.height as f64;
+
+    let width = (bitmap_width * fixup_factor) as usize;
+    let height = (bitmap_height * fixup_factor) as usize;
+
+    let bitmap_width = bitmap_width as usize;
+    let bitmap_height = bitmap_height as usize;
+
+    let b_buf = &bitmap_glyph.buf;
+    let scaling_factor =
+        (bitmap_width as f32 / width as f32).max(bitmap_height as f32 / height as f32);
+    let advance_step = scaling_factor.ceil() as usize;
+    let mut scaled_buffer = Vec::with_capacity(width * height * 3);
+
+    let mut new_line_index = 0;
+    let mut source_line_index = 0;
+
+    while new_line_index < height {
+        let mut new_column_index = 0;
+        let mut source_column_index = 0;
+
+        while new_column_index < width {
+            let mut r: u32 = 0;
+            let mut g: u32 = 0;
+            let mut b: u32 = 0;
+            let mut pixels_picked: u32 = 0;
+
+            let source_end_line = std::cmp::min(source_line_index + advance_step, bitmap_height);
+            let source_end_column = std::cmp::min(source_column_index + advance_step, bitmap_width);
+
+            let mut source_line_index = source_line_index;
+            while source_line_index < source_end_line {
+                let cur_pixel_index = source_line_index * bitmap_width * 3;
+
+                let mut source_column_index = source_column_index;
+                while source_column_index < source_end_column {
+                    r += b_buf[cur_pixel_index + source_column_index * 3] as u32;
+                    g += b_buf[cur_pixel_index + source_column_index * 3 + 1] as u32;
+                    b += b_buf[cur_pixel_index + source_column_index * 3 + 2] as u32;
+                    source_column_index += 1;
+                    pixels_picked += 1;
+                }
+                source_line_index += 1;
+            }
+
+            if pixels_picked == 0 {
+                scaled_buffer.push(0);
+                scaled_buffer.push(0);
+                scaled_buffer.push(0);
+            } else {
+                scaled_buffer.push((r / pixels_picked) as u8);
+                scaled_buffer.push((g / pixels_picked) as u8);
+                scaled_buffer.push((b / pixels_picked) as u8);
+            }
+
+            source_column_index += advance_step;
+            new_column_index += 1;
+        }
+        source_line_index += advance_step;
+        new_line_index += 1;
+    }
+
+    // This top computation performs better with a scaling algo we use.
```

Comment:
```suggestion
    // This top computation performs better with the scaling algorithm we use.
```

Reply:
I'd also like to know *how* it performs better. I'd assume this is not about performance?

Reply:
No, it's about a glyph position. If you just multiply it by fixup_factor it'll be too tight to the top imho.